httpbakery/browser.go

```diff
@@ -26,16 +32,116 @@ func OpenWebBrowser(url *url.URL) error {
+// to be returned from a discharge request.
+//
+// The visitURL parameter holds a URL that should be
+// visited by the user in a web browser; the waitTokenURL parameter
+// holds a URL that can be long-polled to acquire the resulting
+// discharge token.
+//
+// The waitURL parameter is for legacy clients that don't
+// understand about discharge tokens and should
+// hold a URL that can be long-polled to acquire the
+// discharge macaroon.
+func (wi WebBrowserInteractor) SetInteraction(e *Error, visitURL, waitTokenURL, waitURL string) {
+	e.SetInteraction(wi.Kind(), webBrowserParams{
+		VisitURL:     visitURL,
+		WaitTokenURL: waitTokenURL,
+	})
+	logger.Infof("WebBrowserInteractor.SetInteraction %q %q; methods %v", visitURL, waitURL, e.Info.InteractionMethods)
+	// Set the visit and wait URLs for legacy clients too.
+	e.Info.VisitURL = visitURL
+	e.Info.WaitURL = waitURL
+}
+
+type waitTokenResponse struct {
+	Kind  string
+	Token string
+}
+
+// webBrowserParams holds parameters for the WebBrowserInteraction
+// interaction interaction-required error.
+type webBrowserParams struct {
+	// VisitURL holds the URL to be visited in a web browser.
+	VisitURL string
 
-type webBrowserVisitor struct{}
+	// WaitTokenURL holds a URL that will block on GET
+	// until the browser interaction has completed.
+	// On success, the response is expected to hold a waitTokenResponse
+	// in its body holding the token to be returned from the
+	// Interact method.
+	WaitTokenURL string
+}
+
+var (
+	_ Interactor       = WebBrowserInteractor{}
+	_ LegacyInteractor = WebBrowserInteractor{}
+)
 
-func (webBrowserVisitor) VisitWebPage(ctx context.Context, client *Client, methodURLs map[string]*url.URL) error {
-	u := methodURLs[UserInteractionMethod]
-	if u == nil {
-		return ErrMethodNotSupported
+// Interact implements Interactor.Interact by opening a new web page.
+func (wi WebBrowserInteractor) Interact(ctx context.Context, client *Client, location string, irErr *Error) (*DischargeToken, error) {
+	var p webBrowserParams
+	if err := irErr.InteractionMethod(wi.Kind(), &p); err != nil {
+		logger.Infof("WebBrowserInteractor.Interact; no method %v", err)
+		return nil, errgo.Mask(err, errgo.Is(ErrInteractionMethodNotFound))
 	}
-	return OpenWebBrowser(u)
+	logger.Infof("WebBrowserInteractor.Interact %#v", p)
+	visitURL, err := relativeURL(location, p.VisitURL)
+	if err != nil {
+		return nil, errgo.Notef(err, "cannot make relative visit URL")
+	}
+	waitTokenURL, err := relativeURL(location, p.WaitTokenURL)
+	if err != nil {
+		return nil, errgo.Notef(err, "cannot make relative wait URL")
+	}
+	open := wi.OpenWebBrowser
+	if open == nil {
+		open = OpenWebBrowser
+	}
+	if err := open(visitURL); err != nil {
+		return nil, errgo.Mask(err)
+	}
+	return waitForToken(ctx, client, waitTokenURL)
+}
+
+// waitForToken returns a token from a the waitToken URL
+func waitForToken(ctx context.Context, client *Client, waitTokenURL *url.URL) (*DischargeToken, error) {
```

Comment:
As it's possible to do so. Let's support short polling with a timeout here. This will avoid errors where apache/haproxy decide to time out while we're still waiting.